### ISSUE TITLE
Discussion about the direction of this boilerplate

### ISSUE THREAD

**Issue by maxmantz** [mxstbr/react-boilerplate#600]
I would like to start a discussion about the direction this boilerplate is taking. When I started using it a few months ago, I was very happy with the state of things in general. It featured basically all the technologies recommended in a modern ReactJS / Redux app and its configuration - while already complex - was manageable. 

In the last few months I've been worried about the addition of many new features. Necessary changes, like changing the location of sagas & reducers to make larger apps more manageable has been accompanied by vast additions in complexity. For example, we now have a DLL build which is supposed to speed up webpack build times during development. At least for me this has never been an issue, but it adds another dimension to the process, which makes it even harder to understand it all. The webpack build config has heavily increased in complexity, as well as relying on a beta version of webpack itself. In the issue tracker I can see many new additions already planned which will undoubtedly add to the ever growing complexity. If I want to use the boilerplate in its current state and keep my app up to date with it, I will have to adapt - and ultimately learn - about all of these changes, which is a major distraction for me from actually getting things done. 

While I understand that this is supposed to be the state-of-the-art boilerplate for developing redux apps, I seriously wonder if all of these additions add more trouble than they are worth. Whenever I run into issues like the Windows issues I've described (or the test build failures I've reported previously), it is near impossible for me pinpoint the cause because of the countless additions and features implemented here. Any single one of them - or a combination of these can cause unforeseen problems on different platforms. I also wonder, with the many contributions by different people, if there is a single person I can talk to who has a complete understanding of all the little pieces involved here. I seriously hope this is the case, but I fear I will have to gather the information by backtracking diffs and commits and piecing them together myself - without even knowing if this is a futile task.

So in short I strongly urge you to consider the direction this project is taking and not leave users like me who have been using this boilerplate for a while now completely behind.

**Comment by gihrig** [comment]
It seems the issues you have raised might be handled by:
1. More frequent versioned releases (following semver)
2. Keeping changelog.md updated showing features added by version
3. Adding to docs/general/remove.md for major features as they are added. 

Or perhaps take the opposite approach to major new features,  making them opt-in to a core project. Ideally as npm modules. Not sure how one would go about making things like DLL build or i18n into an npm module... This might be more effort than describing how to remove such features.

Would this meet you needs for a stable, updateable boilerplate?

I realize there would be considerable work required to keep the above in sync. But I think it would meet the needs of wider range of projects and time/skill level of end users.

**Comment by maxmantz** [comment]
This is a good idea. At least providing an upgrade path for existing projects is required from my standpoint. I would like to emphasize that I appreciate the amount of work that has been put in by all of the contributors here and I'm sure they have only the best interests in mind. 

The i18n addition you mention is an interesting one. My app has i18n support, but not with react-intl but another library because I found react-intl's syntax to be cumersome and unintuitive. Implementing it into the boilerplate would add something that is from my point of view not essential to get a modern redux app running. Also the integration of this would add a dependency to `package.json` that is not required when the app you want to build has no need for i18n.

In addition to what you've pointed out I would like the maintainers of this project to look at each contribution and determine the following: is it really *required* or merely *nice-to-have*. I brought up the DLL build because from my point of view it goes down the nice-to-have route, while adding a lot of complexity to the build process.

**Comment by mxstbr** [comment]
First of all, thanks for kicking off this discussion @maxmantz, I'm glad we're having this dicussion!

Let me first respond to your specific points:

> For example, we now have a DLL build which is supposed to speed up webpack build times during development. At least for me this has never been an issue […]

HMR is unusable for even slightly big applications because it's so slow. Many people quite simply can't use it, which is a shame because it's a great technology that enables a great development experience. By utilizing the DLL plugin we can circumvent that and make it viable for even bigger apps, which imho is a great addition. Yes, it adds a bit more complexity, but if you don't want it you can always remove it by following the removal guide!

> At least providing an upgrade path for existing projects is required from my standpoint.

First, this is entirely untrue. Whatever state you start from, _I personally guarantee_ that it's amazing to work with.

The `master` branch is vested, whenever you take anything from there I personally stand for the quality it has. I've let two or three bad commits through recently which was 100% my own responsibility, and I've taken note and _will not let that happen again_. 

There is no need to update this boilerplate. Once you're up-and-running, updating it shouldn't need to happen. If you see a nice feature you want to have, you can always follow the PR, but generally you're very well off with whatever state you started from.

Second, I would **love love love** to provide an upgrade path for existing projects but the thing is that it's quite simply not possible. If you have any idea how to make that happen I would very very happily add it.

This is a boilerplate, a starting point, not a library. Updating isn't a thing that should happen, because _the idea of a boilerplate is that the user changes it_. A library isn't meant to be changed by the user, which is why it can be updated with a simple `npm install`.

>  Keeping changelog.md updated showing features added by version

:+1: Will release a new official version as soon as i18n is done and add the changes to the CHANGELOG.

> Adding to docs/general/remove.md for major features as they are added. 

We've done that for most historic ones, and have done it for all (?) new ones. It's one of our main selling points, and I'm committed to keep this in sync.

Now, I'm only human too, so if you see something that doesn't quite work or is missing please _immediately ping me_ and I'll get it done asap.

> Implementing [react-intl] into the boilerplate would add something that is from my point of view not essential to get a modern redux app running.

I'm not 100% convinced, but I see your point here. Could you chime into #391? Thanks!

----

In general, I agree with you. We've not accepted all features and we'll continue to do so. I also think we should remove a few things that add a lot of complexity see #441, #527 and friends. The thing is just that I haven't had the time to properly work on this for the past weeks, but that's about to change so expect some improvements here soon.

Does that answer all questions/concerns?

**Comment by maxmantz** [comment]
Thank you for the very detailed answer! It certainly does answer a lot of questions. Please forgive me when I misinterpreted the usage of the DLL build as being optional. When it fixes stability issues by all means it is definitely required.

I also understand that an upgrade path for existing projects is unrealistic, especially taking different configurations into account. Anyways, once I'm done fixing those issues I have currently I will be happy to write down my experiences when upgrading configurations based on the state I was on previously.

But therein lies the issue. I haven't paid enough attention to the changes in the boilerplate on the config side - which is a mistake I guess. I realised that you had made a few changes to the structure of the app since then - moving sagas and reducers is what intrigued me so I tried to take that into account. But trying to implement that into my existing structure involved injecting sagas into the store, which was not yet possible from my config.

Trying to implement that I also realized that the webpack configuration has changed significantly since then. This opened up another lengthy research about what had changed and what was required. Then when I look through the PRs and issue tracker I am a little scared of the things to come.

The experience of trying to take something as (seemingly) simple like rearranging sagas and selectors and turning it into a very long long journey through code comparisons and diffs was simply a little frustrating. I understand this isn't a library. The configuration is very sophisticated - but therein lies the issue when something as small as this changes - you're basically fighting an uphill battle at that point. I'd once again be very happy to assist in my findings about the issues I currently have to an FAQ if you wish.

**Comment by mxstbr** [comment]
> I'd once again be very happy to assist in my findings about the issues I currently have to an FAQ if you wish.

Absolutely, please do! And if you do come up with a better way we could provide an upgrade path please please let me know!
